macros/src/lib.rs

```diff
@@ -0,0 +1,83 @@
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
+use std::iter::FromIterator;
+
+#[proc_macro_attribute]
+pub fn type_code(args: TokenStream, input: TokenStream) -> TokenStream {
+    let args = syn::parse_macro_input!(args as syn::AttributeArgs);
+    let input = syn::parse_macro_input!(input as syn::ItemEnum);
+
+    let bits = &args[0];
+    let name = &input.ident;
+    let mut variants = Vec::new();
+    let mut decodes = Vec::new();
+    let mut encodes = Vec::new();
+
+    for index in 0..input.variants.len() {
+        let camel = &input.variants[index].ident;
+        let camel_name = camel.to_string();
+
+        if camel_name != "not_used" {
```

Comment:
I think an arguably more clean approach here would be to allow for any unused variant to be marked as such (using `#[not_used]` above that variant) instead of the convention of `not_used` as the variant name to signify this.

Reply:
Yeah... in the well-known case there's a variant called "Permission" that's not used but there are other cases where there is no name given to the unused variants in the spec. Kinda messy but not sure that coming up with placeholder names would be much better, but agree in general that an attribute would be cleaner.